### PR TITLE
lnurl-auth

### PR DESCRIPTION
**Pull request by pseudozach**
# Description

Relates to issue: **ZEUS-0000**

This PR introduces lnurl-auth feature.

This pull request is categorized as a:

- [x] New feature
- [ ] Bug fix
- [ ] Code refactor
- [ ] Configuration change
- [ ] Locales update
- [ ] Other

## Checklist
- [X] I’ve run `npm run tsc` and make sure my code didn’t produce errors 
- [X] I’ve run `npm run prettier` and formatted my code correctly

## Testing

If you added new functionality or fixed a bug, did you add new unit tests?

- [X] No, I’m a fool
- [ ] Yes
- [ ] N/A

I have tested this PR on the following platforms (please specify OS version and phone model/VM):

- [X] Android - Nexus 6 - 7.1.1
- [ ] iOS

I have tested this PR with the following types of nodes (please specify node version and API version where appropriate):

- [X] LND
- [ ] c-lightning-REST
- [ ] spark
- [ ] Eclair
- [ ] LndHub

### Locales
- [X] I’ve added new locale text that requires translations
- [ ] I’m aware that new translations should be made on the Zeus [Transfix page](https://www.transifex.com/ZeusLN/zeus/) and not directly to this repo

### Third Party Dependencies and Packages

- [X] Contributors will need to run `npm install` after this PR is merged in
- [ ] 3rd party dependencies have been modified:
    * verify that `package.json` and `package-lock.json` have been properly updated
    * verify that dependencies are installed for both iOS and Android platforms

### Other:

- [ ] Changes were made that require an update to the README
- [ ] Changes were made that require an update to onboarding

### CODE REVIEW

**Comment by kaloudis** [comment]
What happened with the big change with the `package-lock.json`?

File: views/LnurlAuth.tsx [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+import { Button, Header, Icon } from 'react-native-elements';
+import querystring from 'querystring-es3';
+import url from 'url';
+import SettingsStore from './../stores/SettingsStore';
+import { localeString } from './../utils/LocaleUtils';
+import RESTUtils from './../utils/RESTUtils';
+import Base64Utils from './../utils/Base64Utils';
+import { Hash as sha256Hash, HMAC as sha256HMAC } from 'fast-sha256';
+var EC = require('elliptic').ec;
+var ec = new EC('secp256k1');
```
Comment (kaloudis) [inline views/LnurlAuth.tsx:14]: if these are unchanged, turn them to `const`, otherwise use `let`

**Comment by pseudozach** [comment]
> What happened with the big change with the `package-lock.json`?

I've tried many libraries to get this working on react-native (you can check the initial commit for details), in the end I've removed all unused ones.

File: views/LnurlAuth.tsx (hunk 1 above)
Comment (pseudozach) [reply views/LnurlAuth.tsx]: done

File: utils/Base64Utils.ts [hunk 2]
```diff
@@ -71,6 +71,20 @@ class Base64Utils {
...
+    hexToUint8Array = (hexString: string) => {
+        return new Uint8Array(
+            hexString.match(/.{1,2}/g)!.map(byte => parseInt(byte, 16))
+        );
+    };
+    bytesToHexString = bytes => {
+        return bytes.reduce(function(memo, i) {
+            return memo + ('0' + i.toString(16)).slice(-2); //padd with leading 0 if <16
+        }, '');
+    };
```
Comment (kaloudis) [inline utils/Base64Utils.ts:87]: You can get rid of all the returns and brackets here and make them shorter just like `hexToBase64`

File: views/LnurlAuth.tsx [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+                        )}
+                        {this.state.lnurlAuthSuccess && (
+                            <Text style={{ color: 'green' }}>
+                                {localeString('views.LnurlAuth.loginSuccess')}
+                            </Text>
+                        )}
+                        {!this.state.preparingSignature &&
+                            !this.state.signatureSuccess &&
+                            !this.state.authenticating &&
+                            !this.state.lnurlAuthSuccess && (
```
Comment (kaloudis) [inline views/LnurlAuth.tsx:278]: please destructure these up top, especially if you're using them two or more times

File: views/LnurlAuth.tsx [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+        u.search = querystring.stringify(qs);
+        u.query = querystring.stringify(qs);
+
+        RNFetchBlob.fetch('get', url.format(u))
+            .then((response: any) => {
+                try {
+                    const data = response.json();
+                    return data;
+                } catch (err) {
+                    this.setState({ signatureSuccess: false });
```
[the remark this answers is not in the document]
Comment (pseudozach) [reply views/LnurlAuth.tsx]: good point, done.

File: views/LnurlAuth.tsx [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+    triggerSign() {
+        this.setState({ preparingSignature: true });
+
+        const body = {
+            msg: Base64Utils.btoa(LNURLAUTH_CANONICAL_PHRASE),
+            key_loc: {
+                key_family: 0,
+                key_index: 0
+            }
+        };
```
Comment (kaloudis) [inline views/LnurlAuth.tsx:99]: did you make changes to the `messageSignStore`? `signMessage` just asks for the unencoded string. Open to change that - it just doesn't line up right now

Comment (kaloudis) [reply views/LnurlAuth.tsx:99]: Oh, I see, you're calling the RESTUtils function directly. This is fine.

File: views/LnurlAuth.tsx (hunk 3 above)
Comment (kaloudis) [reply views/LnurlAuth.tsx:285]: line 207

File: views/LnurlAuth.tsx (hunk 5 above)
Comment (pseudozach) [reply views/LnurlAuth.tsx:99]: Yes I just used it the same way.

File: views/LnurlAuth.tsx (hunk 3 above)
Comment (pseudozach) [reply views/LnurlAuth.tsx:285]: You mean line 277?
Can you explain what you mean by destructure?

Comment (pseudozach) [reply views/LnurlAuth.tsx]: done, please take a look

**Review by kaloudis — APPROVED** [review]

**Comment by hsjoberg** [comment]
It would be nice to check interoperatilty between Zeus with lnd backend and Blixt Wallet (uses lnd).

The same seed in both Zeus lnd and Blixt should create the same login on for example http://lnmarkets.com

**Comment by ZeusLN** [comment]
Merging. Will follow up on cross compatibility with Blixt but I have a feeling it will work.

**Comment by pseudozach** [comment]
> It would be nice to check interoperability between Zeus with lnd backend and Blixt Wallet (uses lnd).
> 
> The same seed in both Zeus lnd and Blixt should create the same login on for example https://lnmarkets.com

Sorry it took me this long to spin up a new lnd with a known seed and then use the same seed on a blixt but I finally managed to test, result is surprising to me, they don't seem to match.
But not sure if it's related to lnurl-auth as even the same signed message do not match on these wallets.

Note that blixt is on the mainnet vs lnd is on testnet, would that explain the difference? 

Tested on [lnurl.fiatjaf.com](lnurl.fiatjaf.com)
Here's the LND + Zeus lnurl-auth results
> key	02dc320450e870df0fc52ec2a77969ddd3ac2fe665aa3ebff8215de9334b0b8cc0
k1	8f94d964365a2fd8117f1acadf63461480e3176e7d4900d363f2fe53f99486ce
sig 3045022100f0c56f8ce60743eabe7fc3ce21caceae51ba75d41f92a7afc3c661de55aa521f02202f8274a3288b11b8fc3b0be1f2c783ad833896135c6fc51b92acb81ece9cc9b5

Blixt result with from same seed
> key	029e072f547752f66415a02146eca9ba4fbd39f5dce42e34e6e8a27582214bbdce
k1	dd386458959cd6d96e79d4292c5edd3ad6d975eb797dda066139d5baf27ad4b5
sig	304402203e2df257d0d484999e4ef6a00a6dee7304e73cb86ceab008dd821a2f556a9b85022033f866930c7fde22b42cd72816fcb7af7c1a68fe1d229041abaf702f77e7172c

the seed used:
---------------BEGIN LND CIPHER SEED---------------
 1. about      2. twin      3. hope       4. year   
 5. cheese     6. old       7. crisp      8. weekend
 9. identify  10. urge     11. brain     12. duck   
13. next      14. maximum  15. essence   16. brass  
17. select    18. bleak    19. omit      20. cook   
21. this      22. scene    23. shoulder  24. balance
---------------END LND CIPHER SEED-----------------

I went ahead and signed the canonical phrase with both wallets and ended up with different signatures as well:
Blixt
> d9dfzmrpcwptxjcgfkdk9atohisngi7czgc7aybz8n6cbi35o596y4ffhs9y5bz15nwgqgc6hiz457em8a1ff8ywqkn6s6f3qdscmzsu

lnbtccli3
> d94eyu5dca4i1gijooxy1d9kccqhs71dp3re5qqpk9npnp5owsgqq43dacicx1td3cattdymuxumj9i34gpu19mc8f9fqug9898jrhd1

**Comment by hsjoberg** [comment]
Okay so I think the reason why the signature differs between Zeus lnd and Blixt Wallet is because of different APIs being in use.

In Blixt `signMessage` from the root RPC (`lnrpc`) is being used: https://api.lightning.community/#signmessage
In Zeus lnd `signMessage` from the `signrpc` is being used: https://api.lightning.community/#signmessage-2

The difference between these two is that `signrpc.signMessage` will derive a private key from a BIP-32 like protocol, while `lnrpc.signMessage` will just sign a message with the lightning node's private key.

One of the goals back when lnurl-auth was changed to support the [signing method to derive a "linkingKey"](https://github.com/fiatjaf/lnurl-rfc/blob/master/lnurl-auth.md#linkingkey-derivation-for-wallets-which-dont-have-an-access-to-master-privkey) was to skip using any BIP32 pubkey derivation because lnd does not support full BIP44 4 level derivation anyway (meaning it can only do m'/1/2 and not m'/1/2/3/4 like the lnurl-auth specs previously mandated).

From the lnurl-auth spec:

`2. LN WALLET obtains an RFC6979 deterministic signature of sha256(utf8ToBytes(canonical phrase)) using secp256k1 with node private key.`

So as it stands right now, there won't be any compatibility between Zeus and Blixt regarding lnurl-auth, which in my opinion is a bummer.

My suggestion is to change Zeus lnd to use `lnrpc.signMessage`. I'm not sure if you've already released a version with lnurl-auth support. If yes, as you have a lot of users that might already have started using lnurl-auth, this might not feasible.

Best
Hampus

**Comment by ZeusLN** [comment]
Follow-up PR with compatibility: https://github.com/ZeusLN/zeus/pull/504

**Comment by pseudozach** [comment]
> > It would be nice to check interoperability between Zeus with lnd backend and Blixt Wallet (uses lnd).
> > The same seed in both Zeus lnd and Blixt should create the same login on for example https://lnmarkets.com
> 
...
> 
> the seed used:
> ---------------BEGIN LND CIPHER SEED---------------
> 
> 1. about      2. twin      3. hope       4. year
> 2. cheese     6. old       7. crisp      8. weekend
> 3. identify  10. urge     11. brain     12. duck
> 4. next      14. maximum  15. essence   16. brass
> 5. select    18. bleak    19. omit      20. cook
> 6. this      22. scene    23. shoulder  24. balance
>    ---------------END LND CIPHER SEED-----------------
> 
...

After a looong time testing, I'm happy to announce blixt & zeus (via LND REST) provide the same signature for same phrases.
TL;DR issue was my testing setup: blixt was on mainnet, LND was on testnet which seems to lead to different private keys and different signatures.

Summary:
Restoring above seed in both blixt and LND mainnet nodes and signing below data leads to same results:
sign phrase: test
blixt: d7t8phbaipq4gwwytorme38wtyuoqim9xqcasbb9u6thut39bt7y4wa7tn7k8a83x7rax4x9w3m7pzzxpp6zaddmmwzixh7e8pg5tybd
lnd cli: d7t8phbaipq4gwwytorme38wtyuoqim9xqcasbb9u6thut39bt7y4wa7tn7k8a83x7rax4x9w3m7pzzxpp6zaddmmwzixh7e8pg5tybd
zeus (via lnd rest):
d7t8phbaipq4gwwytorme38wtyuoqim9xqcasbb9u6thut39bt7y4wa7tn7k8a83x7rax4x9w3m7pzzxpp6zaddmmwzixh7e8pg5tybd

sign phrase: [lnurl-auth canonical phrase](https://github.com/fiatjaf/lnurl-rfc/blob/luds/13.md)
blixt:
d9dfzmrpcwptxjcgfkdk9atohisngi7czgc7aybz8n6cbi35o596y4ffhs9y5bz15nwgqgc6hiz457em8a1ff8ywqkn6s6f3qdscmzsu
lnd cli: (note that phrase needs to be in single or double quotes)
d9dfzmrpcwptxjcgfkdk9atohisngi7czgc7aybz8n6cbi35o596y4ffhs9y5bz15nwgqgc6hiz457em8a1ff8ywqkn6s6f3qdscmzsu
zeus (via lnd rest):
d9dfzmrpcwptxjcgfkdk9atohisngi7czgc7aybz8n6cbi35o596y4ffhs9y5bz15nwgqgc6hiz457em8a1ff8ywqkn6s6f3qdscmzsu

Finally, I've tested lnurl-auth log in on https://lnurl.fiatjaf.com

blixt result:
>**key	029e072f547752f66415a02146eca9ba4fbd39f5dce42e34e6e8a27582214bbdce**
k1	08000d05934aee41cb52268206036337a127f5b062348019fad35532cb6037ae
sig	3045022100cbad58ec610e1bee50a36ea82eab63771224e2314b6bfba2929d7964e00cbf1102205f6a010320e6de88ee57d3ce6cc41f3277d68ab36379451b7d1218d4094a3462

zeus result:
>**key	029e072f547752f66415a02146eca9ba4fbd39f5dce42e34e6e8a27582214bbdce**
k1	459491b279022b8a613af89ac7854e8ad678c25a71ac746694220e41a3599b99
sig	3045022100f46ebf138816528ad7a44a418883ee500ab9b2972eec945849ccff0cc318407502205dea7b9c2ec22eec29829523802cac181c10a668ca5a43641753d2c9246f4664

which proves we have lnurl-auth portability between blixt & zeus! 

Thanks for your patience and support,
